agent/csi/plugin.go

```diff
@@ -2,43 +2,179 @@ package csi
+
+const TargetPublishPath string = "/var/lib/docker/publish/%s"
+
+func NewNodePlugin(name string, nodeID string) *NodePlugin {
+	return &NodePlugin{
+		name:      name,
+		nodeID:    nodeID,
+		volumeMap: make(map[string]*volumePublishStatus),
+	}
+}
 
-	// NodeClient is the node service client
-	NodeClient csi.NodeClient
+// GetPublishedPath returns the path at which the provided volume ID is published.
+// Returns an empty string if the volume does not exist.
+func (np *NodePlugin) GetPublishedPath(volumeID string) (string, error) {
+	if volInfo, ok := np.volumeMap[volumeID]; ok {
+		if volInfo.isPublished {
+			return volInfo.publishedPath, nil
+		}
+	}
+	return "", nil
 }
 
 func (np *NodePlugin) NodeGetInfo(ctx context.Context) (*api.NodeCSIInfo, error) {
+	np.mu.RLock()
+	defer np.mu.RUnlock()
 	resp := &csi.NodeGetInfoResponse{
-		NodeId: np.NodeID,
+		NodeId: np.nodeID,
 	}
 
 	return makeNodeInfo(resp), nil
 }
+
+func (np *NodePlugin) NodeStageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+	stagingTarget := fmt.Sprintf(TargetStagePath, volID)
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.mu.Lock()
+	defer np.mu.Unlock()
+
+	v := &volumePublishStatus{
+		stagingPath: stagingTarget,
+	}
+
+	np.volumeMap[volID] = v
+
+	return nil
+}
+
+func (np *NodePlugin) NodeUnstageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.mu.Lock()
+	defer np.mu.Unlock()
+	if v, ok := np.volumeMap[volID]; ok {
+		if v.isPublished {
+			return status.Errorf(codes.InvalidArgument, "VolumeID %s is not unpublished", volID)
+		}
+		delete(np.volumeMap, volID)
+		return nil
+	}
+
+	return status.Errorf(codes.FailedPrecondition, "VolumeID %s is not staged", volID)
+}
+
+func (np *NodePlugin) NodePublishVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.mu.Lock()
+	defer np.mu.Unlock()
+	if v, ok := np.volumeMap[volID]; ok {
+		v.publishedPath = fmt.Sprintf(TargetPublishPath, volID)
+		v.isPublished = true
+		return nil
+	}
+
+	return status.Errorf(codes.FailedPrecondition, "VolumeID %s is not staged", volID)
+}
+
+func (np *NodePlugin) NodeUnpublishVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.mu.Lock()
+	defer np.mu.Unlock()
+	if v, ok := np.volumeMap[volID]; ok {
+		v.publishedPath = ""
+		v.isPublished = false
+		return nil
+	}
+
+	return status.Errorf(codes.FailedPrecondition, "VolumeID %s is not staged", volID)
```

Comment:
s/staged/published/

Reply:
The variable in question is VolumeID, not Volume. Thus, avoiding reference to TargetStagePath.